### ISSUE TITLE
Abnormal latency behavior in csharp helloworld test

### ISSUE THREAD

**Issue by goldenbull** [grpc/grpc#4503]
I'm trying the helloworld example. I write a simple infinite loop in the client as follows:

```
Console.WriteLine("{0}: connecting to server", DateTime.Now.ToString("HH:mm:ss.fffffff"));
var channel = new Channel("127.0.0.1:" + Port, ChannelCredentials.Insecure);
var client = Greeter.NewClient(channel);
Console.WriteLine("{0}: connected to server", DateTime.Now.ToString("HH:mm:ss.fffffff"));
int seq = 0;
while (true) // infinite loop
{
    var sw = Stopwatch.StartNew();
    var reply = client.SayHello(new HelloRequest {Name = "request " + seq++});
    if (sw.ElapsedMilliseconds > 100)
    {
        Console.WriteLine("{0}: Slow!!! {1,5}ms to recv reply: {2}",
                            DateTime.Now.ToString("HH:mm:ss.fffffff"), sw.ElapsedMilliseconds, reply);
    }
}
```

First I started the server, then I started several clients, then close some of them, then started several other clients, and so on. Finally, I got abnormal output: client occasionally has a long latency when invoke the SayHello method as long as 1 or 2 seconds.
The following screen snapshot shows two clients running simultaneously, each has some long-blocked responses. When there is only 1 client, almost every request is blocked for 1 second.
My environment is vs2013 and windows10 x64.

![grpc-csharp-snapshot](https://cloud.githubusercontent.com/assets/9129006/11863669/a9e20422-a4d0-11e5-9bb0-31bf830501c9.png)

**Comment by jtattermusch** [comment]
I will look into this soon.

**Comment by goldenbull** [comment]
Just FYI: I see release-0.12 is available on nuget.org. I created a new project from scratch and downloaded everything from nuget, this issue still exists.

**Comment by jtattermusch** [comment]
@goldenbull  thanks for the update.

**Comment by mleenhardt** [comment]
I was about to log an issue after observing the exact same behavior when trying to benchmark Wcf vs gRPC.

Using Benchmark.Grpc.* from that [repo](https://github.com/mleenhardt/rpc-benchmark), this is what I'm getting:

![image](https://cloud.githubusercontent.com/assets/8008842/12816333/1f4501a0-cb4c-11e5-94fc-497554268994.png)

![image](https://cloud.githubusercontent.com/assets/8008842/12816341/37c2bce0-cb4c-11e5-8bca-f4f66472c6ae.png)

**Comment by jtattermusch** [comment]
Another complaint:

*Hello,
I build simple GRPC server in C# and noticed that each client call take about 1 sec. I tried example (Greeter) with the same result, most cases (99%) grpc server responds in 1 sec. Similar code in C++ works in about 50ms
But C++ client to that C# server shows same 1sec response time. Looks like something is wrong with server.
I am really upset. Anybody has any idea what is going on and why it is ?*

**Comment by ctiller** [comment]
I'd like to spend some time on this. Is there a readily available repro?

**Comment by jtattermusch** [comment]
How to repro:

get upstream/master
build grpc_csharp_ext.sln
build Grpc.sln

go to Grpc.Core.Tests/PerformanceTest.cs

Unignore test UnaryCallPerformance.

Run it, you will see the almost-exactly 1s lags as desribed.

Outputs I got (in millis):
```
985
982
1000
998
1000
1000
1000
999
1
998
999
1000
999
999
999
1000
1
1
1
1
```

```
Stopwatch s = Stopwatch.StartNew();
Calls.BlockingUnaryCall(callDetails, "ABC");
s.Stop();
Console.WriteLine(s.ElapsedMilliseconds);
```

the test will also give you the latency_trace_csharp.txt file to be analyzed by our perf analyzer python script.

**Comment by jtattermusch** [comment]
Ok, it seems that the pluck is taking 1s

```
{"t": 9.006485993, "thd":"unknown", "type": "{", "tag": "CompletionQueueSafeHandle.Pluck", "file": "unknown", "line": 0, "imp": 0}
{"t": 10.005408202, "thd":"unknown", "type": "}", "tag": "CompletionQueueSafeHandle.Pluck", "file": "unknown", "line": 0, "imp": 0}
```

**Comment by jtattermusch** [comment]
Based on yesterday's complaint, it seems like that the client is doing alright, but the server takes 1s to respond (which is why the pluck is taking 1s).

**Comment by ctiller** [comment]
I just sent out #5180 to address this.
We were missing a wakeup in the polling path.

**Comment by jtattermusch** [comment]
The change will be in upcoming 0.13 release (currently merged to release-0_13 branch and will soon appear in master branch as well).

**Comment by mleenhardt** [comment]
Confirming this is resolved in 0.13

![image](https://cloud.githubusercontent.com/assets/8008842/13029556/b178c0b2-d28f-11e5-9fce-a80205bb20c6.png)
